File: java-demo/src/singleton/testSingleton4/singleton4.java
package singleton.testSingleton4;

/**
 * Created by linshaoqin on 12/11/22
 *
 * @author linshaoqin@bytedance.com
 */
public class singleton4 {
    private volatile static singleton4 instance;
    private singleton4(){}
    public static singleton4 getInstance(){
        if(instance==null){
            synchronized (singleton4.class){
                if(instance==null){
                    instance=new singleton4();
                }
            }
        }
        return instance;
    }
}

File: java-demo/src/singleton/testSingleton4/singleton4.kt
package singleton.testSingleton4

/**
 * Created by linshaoqin on 12/11/22
 * @author linshaoqin@bytedance.com
 */
class singleton4kt private constructor(){
    companion object {
//        val instance: singleton4kt by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) {
//            singleton4kt()
//        }

//        val instance: Lazy<singleton4kt>
//            get() {
//                return lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) {
//                    singleton4kt()
//                }
//            }
    }
}